lib/spack/spack/test/packaging.py

```diff
@@ -0,0 +1,252 @@
+##############################################################################
+# Copyright (c) 2013-2016, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the NOTICE and LICENSE files for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+"""
+This test checks the binary packaging infrastructure
+"""
+import pytest
+import spack
+import spack.store
+from spack.fetch_strategy import URLFetchStrategy, FetchStrategyComposite
+from spack.spec import Spec
+import spack.binary_distribution as bindist
+from llnl.util.filesystem import join_path, mkdirp
+import argparse
+import spack.cmd.buildcache as buildcache
+from spack.relocate import *
+import os
+import stat
+import sys
+import shutil
+
+
+@pytest.fixture(scope='function')
+def testing_gpg_directory(tmpdir):
+    old_gpg_path = spack.util.gpg.GNUPGHOME
+    spack.util.gpg.GNUPGHOME = str(tmpdir.join('gpg'))
+    yield
+    spack.util.gpg.GNUPGHOME = old_gpg_path
+
+
+def has_gnupg2():
+    try:
+        spack.util.gpg.Gpg.gpg()('--version', output=os.devnull)
+        return True
+    except Exception:
+        return False
+
+
+def fake_fetchify(url, pkg):
+    """Fake the URL for a package so it downloads from a file."""
+    fetcher = FetchStrategyComposite()
+    fetcher.append(URLFetchStrategy(url))
+    pkg.fetcher = fetcher
+
+
+@pytest.mark.usefixtures('install_mockery', 'testing_gpg_directory')
+def test_packaging(mock_archive, tmpdir):
```

Comment:
This could use more validation. For example: when you create a build cache and add packages to it, are they actually there? Generally I would like to see this split up more too, but IMO that's not a priority.